avocado/core/job.py

```diff
@@ -556,6 +563,8 @@ def run(self):
             self.log.error('Report bugs visiting %s', _NEW_ISSUE_LINK)
             return exit_codes.AVOCADO_FAIL
         finally:
+            if self.job_post_dispatcher.extensions:
+                self.job_post_dispatcher.map_method('run', self)
```

Comment:
I disagree for a number of reasons:

1) The current plugin interfaces are that same granularity, look at `avocado/plugins/base.py:CLI` which allows for new command line *options*. We could have added interfaces for also defining new commands, but it was decided to keep things separate. It's cleaner IMHO.

2) Since we're defining mandatory interfaces (using abstract base classes) this would require the user to implement `pre` even if he's only interest on `post`. The email notification is one example, the user may be interested only in job completion notifications. Also, a breakage of the original design.

3) If the pre/post plugin are similar but have to respond based on the "event" type, this can be coded in numerous ways:

```python
import time
import logging

from avocado.core.settings import settings
from avocado.plugins.base import Job


class SleepBase(JobPrePost):

    name = 'sleep'
    description = 'Sleeps for a number of seconds'
    event = None

    def run(self, job):
        log = logging.getLogger("avocado.app")
        seconds = settings.get_value(section="plugins.job.sleep",
                                     key=event,
                                     key_type=int,
                                     default=3)
        for i in xrange(1, seconds + 1):
            log.info("Sleeping %s job %2i/%s", self.event, i, seconds)
            time.sleep(1)

class SleepPre(SleepBase):
    event = 'pre'

class SleepPost(SleepBase):
    event = 'post'
```